segment_store/src/segment.rs

```diff
@@ -736,16 +751,65 @@ impl Segment {
         }
     }
 
-    // Optimised read group method where only a single column is being used as
-    // the group key.
-    //
-    // In this case the groups can be represented by a single integer key.
-    fn read_group_single_group_column(
-        &self,
-        predicates: &[Predicate<'_>],
-        dst: &mut ReadGroupResult<'_>,
+    // Optimised read_group path for queries where only a single column is being
+    // grouped on. In this case building a hash table is not necessary, and
+    // the group keys can be used as indexes into a vector whose values contain
+    // aggregates. As rows are processed these aggregates can be updated in
+    // constant time.
+    fn read_group_single_group_column<'a>(
+        &'a self,
+        dst: &mut ReadGroupResult<'a>,
+        groupby_encoded_ids: &[u32],
+        aggregate_columns_data: Vec<Values<'a>>,
     ) {
-        todo!()
+        let column = self.column_by_name(dst.group_columns[0]);
+        assert_eq!(dst.group_columns.len(), aggregate_columns_data.len());
+        let total_rows = groupby_encoded_ids.len();
+
+        // Allocate a vector to hold aggregates that can be updated as rows are
+        // processed. An extra group is required because encoded ids are 0-indexed.
+        let required_groups = groupby_encoded_ids.iter().max().unwrap() + 1;
+        let mut groups: Vec<Option<Vec<AggregateResult<'_>>>> =
+            vec![None; required_groups as usize];
+
+        for (row, encoded_id) in groupby_encoded_ids.iter().enumerate() {
+            let idx = *encoded_id as usize;
+            match &mut groups[idx] {
+                Some(group_key_aggs) => {
+                    // Update all aggregates for the group key
+                    for (i, values) in aggregate_columns_data.iter().enumerate() {
+                        group_key_aggs[i].update(values.value(row));
+                    }
+                }
+                None => {
+                    let mut group_key_aggs = Vec::with_capacity(dst.aggregate_columns.len());
+                    for (_, agg_type) in &dst.aggregate_columns {
+                        group_key_aggs.push(AggregateResult::from(agg_type));
+                    }
```

Comment:
Maybe this could be written in a more functional style (proposal is untested): 
```suggestion
                    let mut group_key_aggs = dst
                      .aggregate_columns
                      .iter()
                      .map(|(_, agg_type) AggregateResult::from(agg_type))
                      .collect::<Vec<_>>()
```

Reply:
Ah yes. This is a bad habit on my part and I did something similar (and not idiomatic) on a previous branch. Good change.